staging/src/k8s.io/apiserver/pkg/storage/cacher/caching_object.go

```diff
@@ -0,0 +1,174 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cacher
+
+import (
+	"bytes"
+	"io"
+	"sync"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+// FIXME:
+type serializationResult struct {
+	once sync.Once
+
+	// raw is serializaed object.
+	raw []byte
+	// err is error from serialization.
+	err error
+}
+
+// FIXME:
+type CachingVersionedObject struct {
+	// FIXME:
+	Object runtime.Object
+
+	// FIXME: ???
+	lock sync.Mutex
+	// FIXME: sync.Map looks like a good usecase for us:
+	// - StoreAndLoad a function that computes stuff
+	// - the function that computes has sync.Once underneath.
+	serialization map[runtime.Encoder]*serializationResult
+}
+
+// FIXME:
+func (o *CachingVersionedObject) Encode(e runtime.WithVersionEncoder, w io.Writer) error {
+	serialization := func() *serializationResult {
+		// FIXME: It probably would benefit to have a fast path,
+		//   instead of always locking.
+
+		o.lock.Lock()
+		defer o.lock.Unlock()
+
+		serialization, exists := o.serialization[e.Encoder]
+		if exists {
+			return serialization
+		}
+
+		serialization = &serializationResult{}
+		o.serialization[e.Encoder] = serialization
+		return serialization
+	}()
+	serialization.once.Do(func() {
+		buffer := bytes.NewBuffer(nil)
+		serialization.err = e.Encoder.Encode(o.Object.DeepCopyObject(), buffer)
+		serialization.raw = buffer.Bytes()
+	})
+	// Once invoked, fields of serialization will not change.
+	if serialization.err != nil {
+		return serialization.err
+	}
+	_, err := w.Write(serialization.raw)
+	return err
+}
+
+func (o *CachingVersionedObject) GetObjectKind() schema.ObjectKind {
+	return o.Object.GetObjectKind()
+}
+
+func (o *CachingVersionedObject) DeepCopyObject() runtime.Object {
+	// FIXME:
+	panic("CachingVersionedObject should never be deep-copied")
+	return o
+}
+
+// FIXME: Add comment.
+//
+// FIXME: This should implement metav1.Object interface probably.
```

Comment:
Idea: implement the mutating parts of the metadata accessor and panic, to try and help ensure that nothing mutates this?

Reply:
i had very similar idea - but given I know how to do that and this was supposed to be a POC that this is doable, I didn't yet do that.

So yes - I agree we should do that.